### PR TITLE
Dynamo Export use fake tensor

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94276

This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
> These changes look... plausible, I guess? Is there any test?

Yes, test_dynamic_shapes_right_side in test/dynamo/test_repros.py covers this case.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -642,22 +646,28 @@ def output(self, target, args, kwargs):
...
+                if isinstance(input, FakeTensor):
+                    fake_mode = input.fake_mode
+                    break
+
+        with fake_mode:
+            graph = make_fx(
+                graph_with_interpreter,
+                decomposition_table=decomposition_table,
+                _allow_non_fake_inputs=True,
+            )(*retrace_args)
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:670]: I don't really understand this. Deferring to @voznesenskym on the review here

**Comment by ezyang** [comment]
I don't feel like the eager side version of fine grained dynamic shapes should depend on this PR. I get the feeling this PR may take a while to review.

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -642,22 +646,28 @@ def output(self, target, args, kwargs):
...
-        )(*graph_captured_input)
+        retrace_args = graph_captured_input
+        fake_mode = null_context()
+
+        if tracing_mode == "symbolic":
+            retrace_args = fakified_example_inputs
+            for input in fakified_example_inputs:   # type: ignore[union-attr]
+                if isinstance(input, FakeTensor):
+                    fake_mode = input.fake_mode
+                    break
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:663]: use the util, fake_mode_from_tensors

File: torch/_dynamo/eval_frame.py [hunk 3]
```diff
@@ -575,9 +577,11 @@ def dynamo_normalization_capturing_compiler(
         gm: torch.fx.GraphModule, example_inputs
     ):
         nonlocal graph
+        nonlocal fakified_example_inputs
 
         assert graph is None, "whole graph export entails exactly one graph"
         graph = gm
+        fakified_example_inputs = example_inputs
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:584]: These are compile time inputs, if you want to extract them, please call them that.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
This was hard to follow. (1) Please refactor this to indicate that you want `symbolic` to use compile_time args. (2) Is the fake mode logic for symbolic redundant with make_fx?

File: torch/_dynamo/eval_frame.py (hunk 3 above)
Comment (SherlockNoMad) [reply torch/_dynamo/eval_frame.py:584]: renamed to compile_time_inputs

**Comment by SherlockNoMad** [comment]
> This was hard to follow. (1) Please refactor this to indicate that you want `symbolic` to use compile_time args. (2) Is the fake mode logic for symbolic redundant with make_fx?

1) Done. 
2) FakeMode logic for symbolic is not redundant for make_fx, as it's calling make_fx with tracing_mode="real". I tried to fix make_fx, to use FakeTensorMode when inputs is a fake tensor, regardless the tracing_mode spplied, but I couldn't get it to work.

File: torch/_dynamo/eval_frame.py [hunk 4]
```diff
@@ -631,22 +634,28 @@ def output(self, target, args, kwargs):
...
+            )(*graph_captured_input)
+        elif tracing_mode == "symbolic":
+            # For dynamic shape, we need to make_fx through the graph with fake tensors under FakeTensorMode
+            # The fake tensors may contain the fine grain dynamic shape passed down from dynamo
+            fake_mode = fake_mode_from_tensors(compile_time_inputs)
+            with fake_mode:
+                graph = make_fx(
+                    graph_with_interpreter,
+                    decomposition_table=decomposition_table,
+                )(*compile_time_inputs)
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:656]: Can fake mode be a requirement for symbolic in make_fx? Can we push this down?

File: torch/fx/interpreter.py [hunk 5]
```diff
@@ -153,6 +153,7 @@ def run(self, *args, initial_env : Optional[Dict[Node, Any]] = None, enable_io_p
 
     @contextmanager
     def _set_current_node(self, node):
+        self.current_node = node
```
Comment (voznesenskym) [inline torch/fx/interpreter.py:156]: why?

File: torch/_dynamo/eval_frame.py (hunk 4 above)
Comment (SherlockNoMad) [reply torch/_dynamo/eval_frame.py:656]: Notice that, this is calling make_fx with "real" tracing mode using FakeTensor, not "symbolic" tracing mode with FakeTensor. 

I tried pushing FakeMode into make_fx, but I couldn't get it passing the UTs. 
Here's what I have tried: in make_fx, if the tracing model is "real" and if any of the input tensor is a FakeTensor, I set the fakemode.

File: torch/fx/interpreter.py (hunk 5 above)
Comment (SherlockNoMad) [reply torch/fx/interpreter.py:156]: ChangeInputOutputSignature uses self.current_node. 

Alos, having self.current_node available is generally useful.

Comment (voznesenskym) [reply torch/fx/interpreter.py:156]: Nit, kinda unrelated to this PR. sounds fine otherwise!

File: torch/_dynamo/eval_frame.py (hunk 4 above)
Comment (SherlockNoMad) [reply torch/_dynamo/eval_frame.py:656]: This is what we do today in aot_autograd. Inside aotautograd, we always ignore user specified tracing mode, and always use the tracing_mode="real". According to @Chillee, it's also expected to use tracing_mode="real" internally. 

See
https://github.com/pytorch/pytorch/blob/master/torch/_functorch/aot_autograd.py#L1053
and
https://github.com/pytorch/pytorch/blob/master/torch/_functorch/aot_autograd.py#L1643

I tried changing the behavior of make_fx, but make_fx is used in so many places, and it's breaks bunch of tests, and it appears that there are subtly of how FakeTensor behave differently under FakeTensorMode vs. non-FakeTensorMode. 

So to me, the easiest path forward is to replicate what AOTAutograd is doing in dynamo.export.

**Comment by voznesenskym** [comment]
Code looks good! Please just add some more tests, export is a very heavily tested codepath already, but more wont hurt :) very close to a stamp!

**Comment by voznesenskym** [comment]
Sherlock assured me 1:1 that existing tests cover this.

**Review by voznesenskym — APPROVED** [review]

**Comment by zhxchen17** [comment]
@SherlockNoMad have you tested on internal models?
